Add support for Ruby 3.x -> 4.0, Rails 6.1 -> 8.1

anitagraham opened the pull request:
This PR and its matching PR refinery/refinerycms-i18n#104 allow `refinerycms` to run with Ruby 3 / Ruby 4.

It has been tested with Ruby 3.2 through 4.0, and with Rails 6.1 through 8.1.

We've removed `coffee-rails` as a dependency and converted those to JS.

anitagraham committed:
Tested with Rails 6.1
Ruby 3.1.2
core: TranslationHelper, SiteBarHelper - using Ruby3 keyword args
      lib/refinery/engine uses new mobility configuration
images, resources: Updated examples to pass an explicit block to 'expect' as passing an implicit block is deprecated.
images: Use errors.add to add errors to an object - had already been done for resources.

Gem versions changed:
  mobility 1.2.9
  rspec-rails 6.0.0.rc1
  friendly_id >= 5.4.0
  friendly_id-mobility ~> 1.0.3

anitagraham committed:
Use matching ruby3 branch of refinerycms-i18n while still a work in progress

anitagraham committed:
Change workflow to test only Ruby 3.1.  Correct reference to refinerycms-i18n in Gemfile'

anitagraham committed:
Merge branch 'master' into ruby3

anitagraham commented:
### Tests failing:

I have seen some of these errors during testing, but thought I was past them.  Many seem to be timeouts, and then the reporting of timeouts seems to be a problem too. 

There are many variations of:

```
 1.1) Failure/Error: **visit refinery.admin_images_path**
          RuntimeError:
            Rack application timed out during boot
          # ./core/spec/system/refinery/admin/xhr_paging_spec.rb:19:in `block (3 levels) in <module:Refinery>'
```

and I saw some other reporting of timeouts on `visit`, as a result of changes to `bundler`, I think.

Also frequently seen (below), which seems to be an error reporting  timeouts. It occurs every minute during a run.
```
{"time":"2022-07-18T05:05:32+00:00",
 "severity":"warn",
 "class":"Async::Task",
 "oid":130980,
 "pid":2890,
 "subject":
    "#<Async::Task:0x000000000001ffa4 /home/runner/work/refinerycms/refinerycms/vendor/bundle/ruby/3.1.0/gems/async-2.0.3/lib/async/task.rb:82:in
    `backtrace' (failed)>",
    "message":[
      "Task may have ended with unhandled exception.",
      "undefined method `logger' for Async:Module
					Async.logger.debug (self) {\"Running server...\"}
					     ^^^^^^^"
					     ],
		"error":{
		  "kind":"NoMethodError",
		  "message":"undefined method `logger' for Async:Module
					Async.logger.debug (self) {\"Running server...\"}
					     ^^^^^^^",
     "stack":
      "/home/runner/work/refinerycms/refinerycms/vendor/bundle/ruby/3.1.0/gems/falcon-capybara-1.5.1/lib/falcon/capybara/wrapper.rb:61:in `block in call'
        /home/runner/work/refinerycms/refinerycms/vendor/bundle/ruby/3.1.0/gems/async-2.0.3/lib/async/task.rb:255:in `block in schedule'
    "}
}
```

anitagraham committed:
Use Ruby 3 syntax callling Resource.create.  Add a description for File and Directory matchers.

anitagraham committed:
Tidy up configurations for mobility 1.2.9

Tidy up configuration block for Mobility 1.2.9
Revert some changes for single argument keyword args
Add descriptions for Generator spec matchers (thereby avoiding a very long message when there are no descriptions).

anitagraham committed:
WIP

parndt commented:
Thank you for looking at this

pdornfel commented:
Is there any update on the status of this PR? I need to upgrade to ruby 3.2.1 and rails 6.x because I need to upgrade to the newest cflinux4 buildpack and no previous versions of ruby are supported. I currently use refinerycms with ruby 2.5.5 and rails 5.x. 

I think this merge will help me upgrade to the versions of ruby and rails mentioned above. Thank you.

parndt committed:
Merge branch 'main' into ruby3

parndt commented on .gitignore:
```diff
@@ -97,3 +97,4 @@ Gemfile.lock
 
 # rspec failures
 .rspec_failures
+Refinery Review.md
```
```suggestion
```

parndt commented:
@pdornfel I'd really love to merge this but CI isn't happy. The corresponding PR for refinerycms-i18n is failing on Ruby 3.0 and 3.2, but not 3.1 🤔 https://github.com/refinery/refinerycms-i18n/pull/104

Any help you or anyone else can lend with figuring out why this is happening would be greatly appreciated. Right now, I suspect the [routing-filter](https://github.com/svenfuchs/routing-filter/) gem we use.

anitagraham committed:
Merge pull request #6 from refinery/main

Merge refinery fixes

anitagraham committed:
Merge remote-tracking branch 'origin/ruby3' into ruby3

anitagraham committed:
remove all capybara configuration from spec/spec_helper to already existing file spec/support/capybara.rb.

remove requirement for 'webdrivers/chromedriver' as this is all handled by Selenium

consistent errors in Selenium chrome webdriver, temporarily switched to firefox.

anitagraham committed:
remove reliance on www.example.com in tests, use attribute selectors to target the part of the url which should change.

anitagraham committed:
previous usage of 'Pages' as a constant referring to Refinery::Pages was failing in Refinery::Admin tests, so now using Refinery::Pages(*)

anitagraham committed:
saving generator options before switching to 'pretend=true' inside the options hash, as options[:original_options]

anitagraham committed:
remove another instance of 'www.example.com'

anitagraham committed:
refinery_help_tag will always show the 'tooltip' message

anitagraham committed:
css cleanup: remove units from unitless values. (0px -> 0)

anitagraham committed:
css cleanup: remove recent_activities and recent_inquiries from stylesheets.

anitagraham committed:
use mixins for common styles between image views

anitagraham committed:
define ImageObject for use in Admin image views
return ImageObjects from Refinery::Admin::ImagesController#index action
unify code for grid_view and list_view into _image.html.erb
add view_objects directory to engine autoload_path

anitagraham committed:
cleanup stylesheets:
 change 0px to unitless 0.

anitagraham committed:
use recommended dragonfly call for stripping images instead of deprecated "process.convert"

anitagraham committed:
scss style reformatted all over to add blank lines and spaces.

anitagraham committed:
Add defined colours for image index actions.
extend image mixins
use mixins to handle grid view and list view for images.

anitagraham committed:
Add defined colours for image index actions.
extend image mixins
use mixins to handle grid view and list view for images.
replace hash.merge with Ruby3 keyword format

anitagraham committed:
Change naming of 'grid_view' and 'list_view' to 'grid_format' and 'grid_view'.
(view, format and layout are already special words, and I may yet change format to something useful, but not duplicating a term used in another context)

keep the format variable as a symbol wherever it is used.

renamed configuration accessors to 'index_formats' (et al) to be clear about their function.

anitagraham committed:
Pages

simplify view code for the admin index of a page.

put selectors used in testing into single file (still some way to go)

anitagraham committed:
Pages

move helpers for actions (edit/preview/delete etc) from tag_helper.rb to (new)action_helper.rb
move helpers for actions (edit/preview/delete etc) from tag_helper.rb to (new)action_helper.rb

simplify view code for the admin index of a page.

put selectors used in testing into single file (still some way to go)

anitagraham committed:
refactor(assets): stylesheets

use simple selectors for file_type icons, use IconHelper to match file_type to icon

use grid layout for images grid

anitagraham committed:
refactor(refinery-images): make ruby3 compatible

use data-attributes hash in markup, in place of "data-#{attributes}..."

define image_presenters for admin/index
unify code for grid_view and list_view

presenters:
group_presenter:  handle the formatting for a group of images

list_presenter and grid_presenter inherit from group presenter,

image_presenter:  generate an entry for the image_index, including clickable links, text information, and actions.

configuration:

define admin_image_sizes in configuration

 replace hard coded image sizes from several locations in views and helpers.

rename Refinery::Images.image_views to Refinery::Images.index_views

// note: looking for a better name for these I tried :formats and :layouts before noting that they, like :views, are magic words for Ruby/Rails and are, therefore, no better. //

remove temporary changes to admin/images_controller
remove image view objects (using image presenters)

split grid_view and list_view in the test contexts

anitagraham committed:
Changes to refinery/pages

anitagraham committed:
refactor(refinery-core): new helper 'locales_with_translated_field'

add new helper 'locales_with_translated_field'  in TranslationHelper to replace repeated code block in index templates.

add test for helper

call from index pages to identify locales with existing translations

anitagraham committed:
refactor(refinery-core): new helper 'edit_in_locales'

add new helper 'edit_in_locales' to ActionHelper

call from index pages to offer links to records with translations
add edit_in_locales to images/pages/resources indexes

anitagraham committed:
add resource_helper to resources_controller

add newline at eof for resources/admin/form

anitagraham committed:
Revert "Change naming of 'grid_view' and 'list_view' to 'grid_format' and 'grid_view'. (view, format and layout are already special words, and I may yet change format to something useful, but not duplicating a term used in another context)"

This reverts commit 6eafef5f

anitagraham committed:
update jquery-ui-rails to version '~7.0.0'

anitagraham committed:
update Rails to ~>7.1

anitagraham committed:
update dependencies in refinerycms-testing

anitagraham committed:
stylesheets: refactor locale icons/buttons

define colours for locale icon backgrounds and text

use id attribute of link to determine id

define mixins for locale elements
remove browser-specific selectors from the rounded-corner mixins

anitagraham committed:
refactor locale actions

share code generating a link to switch/set locale

change name from 'locales_with_titles' to 'locales_with_translated_field' to better reflect what it does.

add messages keys for language and locale in en.yml

anitagraham committed:
Changes from ruby3 branch which are really Rails 7 changes

parndt committed:
Merge branch 'main' into ruby3

parndt committed:
Remove Gemfile lines that interfere with CI

parndt committed:
Update Ruby versions and checkout action in workflow

parndt committed:
Update factory_bot_rails dependency version

parndt committed:
Update rspec-rails dependency to version ~> 6.1

parndt committed:
Update rspec-rails gem version in Gemfile

parndt committed:
Add sqlite to database matrix in CI workflow

parndt committed:
allow sqlite3

parndt committed:
Change sqlite to sqlite3 in workflow matrix

parndt committed:
Add decorators gem for Zeitwerk compatibility

Added a workaround for Zeitwerk loading issues by including the decorators gem from a specific GitHub branch.

parndt committed:
Refactor Capybara app configuration for Rack 3+

Update Capybara configuration to handle Rack 3+ compatibility.

parndt committed:
Simplify find actions handling in crud.rb

parndt committed:
Attempt a few patches

parndt committed:
Images needs imagemagick

parndt commented:
@anitagraham just the pages specs fail now!

parndt committed:
Try some dubious fixes

parndt commented:
Happy new year!

parndt committed:
Tidy up dubious fixes

parndt committed:
Tidy up a bit

parndt committed:
refactor(images): extract presenter refactoring to feature branch

This commit extracts all presenter-pattern refactoring from the ruby3
branch to feature/images-presenter-refactor, resetting images/ to match
main's structure. This allows ruby3 branch to focus solely on Ruby 3 and
Rails compatibility fixes.

Extracted to feature/images-presenter-refactor:
- GridPresenter, GroupPresenter, ImagePresenter, ListPresenter classes
- ImageObject view object
- admin_image_sizes configuration
- View template consolidation (_image.html.erb partial)
- Naming changes (index_views → image_views, preferred_index_view, etc.)
- Helper refactoring (other_index_views, localized, thumbnail URL format)
- Spec context refactoring (grid_view, list_view shared contexts)

Retained on ruby3 (will be reapplied in next commit):
- None yet - this commit restores to main's baseline

These changes were bundled in commit 082bc6003 "refactor(refinery-images):
make ruby3 compatible" but represent UI/architecture improvements separate
from Ruby 3 compatibility requirements.

The feature branch tests pass (101 examples, 0 failures). Next commit will
apply only the minimal fixes required for Ruby 3 compatibility.

parndt committed:
fix(images): add minimal Ruby 3 and Rails compatibility fixes

Apply only the essential changes required for Ruby 3 and modern Rails
compatibility, without any UI/architecture refactoring.

Changes:
- Controller: Add frozen_string_literal, require will_paginate/array
- Controller: Add find_actions to crudify (required for updated CRUD)
- Controller: Fix strong params to allow array uploads: { image: [] }
- Helper: Add frozen_string_literal
- Helper: Convert data attributes to Ruby 3 keyword args format
  - Old: :"data-original" => value
  - New: { data: { original: value } }
- Helper: Convert :geometry => to geometry: keyword syntax
- Helper: Add locale_text_icon() stub (flags removed in refinerycms-i18n PR#104)
- View: Update data attributes to use new hash format with splat operator
- View: Use dom_id() helper instead of manual string interpolation
- Validators: Update to use errors.add() instead of deprecated errors[:key] <<
- Validators: Add frozen_string_literal, inline private methods
- Validators: Convert hash rockets to keyword arguments
- Specs: Fix RSpec expect blocks (explicit vs implicit)
  - Old: expect(lambda)
  - New: expect { lambda.call }
- Specs: Fix XPath selector to use contains() for multi-class elements
  - Fixes preview icon detection with multiple CSS classes
- Specs: Update locale picker test to use ID selector instead of text
  - Old: click_link "FR" (text no longer available without flag images)
  - New: find("a#fr").click

Core changes:
- TranslationHelper: Convert t() method to use keyword arguments (**options)
- ActionHelper: Add locale_language() helper to get locale display name
- Config: Restore main's locale_picker partial structure

Test Results:
- 99 examples, 0 failures, 8 pending

All main branch specs now pass with Ruby 3.2.9.

Related to commit 082bc6003 but extracts only compatibility requirements.

parndt committed:
Fix locale picker to display locale codes instead of language names

refinerycms-i18n PR#104 removed flag images, requiring text-based locale
indicators. Changed edit_in_locale helper to display uppercase locale codes
(e.g., "FR", "EN") instead of full language names when in label mode.

This allows the translation spec to test locale switching by clicking the
locale code link text.

parndt committed:
disable selenium telemetry

parndt committed:
Add locales_with_translated_field helper for locale picker views

The refactored pages and resources views use locales_with_translated_field
helper to display locale indicators. This helper queries translations to
find which locales have content for a given field.

Also revert xhr_paging spec to use image_grid selector instead of
image_index, which was part of the UI refactoring that should stay in
the feature branch.

parndt committed:
Fix xhr_paging spec to ensure grid view is used

Use allow() to stub preferred_image_view to :grid, consistent with the
existing mocking pattern for per_page. This ensures the spec tests grid
view without side effects on other tests.

parndt committed:
Use allow() for config stubs instead of setting/unsetting in after blocks

Replace before/after pattern with allow() stubs for:
- Refinery::Pages.auto_expand_admin_tree
- Refinery::I18n.frontend_locales
- Refinery::Pages.new_page_parts

This is more consistent with the existing mocking pattern, automatically
cleans up after each test, and avoids potential side effects from mutating
global configuration state.

parndt commented:
I've extracted the non-ruby-upgrade-related code into [feature/images-presenter-refactor](https://github.com/refinery/refinerycms/tree/feature/images-presenter-refactor) to keep this PR targeted.

parndt committed:
Add specs for flash notice after successful upload

Test that flash notice appears after successfully uploading an image
or resource through the main interface (not dialog). This ensures the
flash message displays the correct title and doesn't persist across
page reloads.

These specs serve as regression tests to prevent flash message issues.

parndt committed:
misc fixes

parndt committed:
bring back image grid

parndt committed:
Fixups to have new things and old things still working

parndt committed:
Test all the ruby/rails versions we can

Try to support the older versions

parndt committed:
Try converting coffee to JS

parndt committed:
For now, precompile assets for Rails ~> 8.1.0

parndt committed:
Reverse order of ruby tests

parndt committed:
Try to remove the logger override

parndt committed:
Try to rescue the dreaded logger error

parndt committed:
Try to update the docs as much as possible

parndt commented:
Thanks so much for all your hard work on this, it's been extremely helpful.

parndt reviewed (approved):
Amazing that we got here.

parndt committed:
Some name/label pedantry

parndt commented on Gemfile:
```diff
@@ -1,34 +1,55 @@
 end
 
-gem 'refinerycms-i18n', github: 'refinery/refinerycms-i18n', branch: 'master'
+gem 'refinerycms-i18n', github: 'anitagraham/refinerycms-i18n', branch: 'ruby3'
```
```suggestion
gem 'refinerycms-i18n', github: 'refinery/refinerycms-i18n', branch: 'main'
```

parndt committed:
Update Gemfile

parndt commented on Gemfile:
```diff
@@ -1,34 +1,55 @@
-  git: 'https://github.com/refinery/refinerycms-acts-as-indexed',
+gem 'refinerycms-acts-as-indexed', '~> 4.0', '>= 4.0.0',
+  github: 'refinery/refinerycms-acts-as-indexed',
   branch: 'master'
```
```suggestion
  branch: 'main'
```

parndt committed:
Update Gemfile
